Feature/post initial process document endpoint

MirelaGeorgiou opened the pull request:
Endpoint mainly developed by Liudvikas.

Endpoint inserts a skeleton of a MAT process document.

- Added endpoint
- Added validation for request and exception types

Duslerke committed:
Created Post Process Request Boundary Object

Duslerke committed:
Helper method to create fake PostInitialProcessDocumentRequest object.

Added a helper method to the tests project that creates a PostInitialProcessDocumentRequest object so that the 'arrange' phase of tests could be shorter.

Duslerke committed:
Added a method overload for creating MatProcessData Domain object

-----Implementation------
Had to extract dateOfCreation out, since I believe that the dates DateCreated and DateLastModified have to be equal. The alternative is them differing by miliseconds, which might or might not cause issues in the future - who knows how it's going to be used.

The process stage... I left it as a string because I was told that it's still undecided on its format. Therefore I don't feel like changing a lot of existing code around that for an uncertainty.

DateCompleted - a minimum value. Not elegant, but it was decided that it's better than having a nullable DateTime type.

Why not an instance method? Was following the previous method in the factory. Don't understand enough C# to question that decision yet. I guess, it's all good as long as this method doesn't hold any state or anything.

------Tests--------
Using reflections because the regular comparison that's commented out returns a nonsensical error saying that it expected <{ }>, but instead got <{ }>.

It might have been simply comparing object references, but who knows.

So instead I check how many fields the anonymous object has. It's supposed to be zero, since the requirement is that it has to be set as empty object.

Duslerke committed:
Added a PostInitialProcessDocument endpoint Usecase.

--------- Tests ----------
Did not include all of the domain object property fields in the mock gateway verify assertion because test doing that is in the Factory tests. Instead I aimed to show that the object passed into the gateway contains some fields from the request object, and some fieds generated from the factory... to show that request object data was mapped onto domain object, as well as to show that it was the factory that did it by including a field that only it generates (that's not a default of some type in C# (like 'int' default is 0)).

This usecase has only a single test, since it's only invoking other methods and not expecting or returning any output.
The existing test could have been split into 3 tests if the we hadn't agreed on making factory methods static. The issue is that static methods can not be mocked, therefore you can't check if there was a call made to a static method.
The way the test could have been split would have been: doing a test to see if usecase calls the factory.
The other one to see if use case calls the mock gateway with the mock factory output.
The 3rd test would be similar to 2nd one except the factory is not a mock any more - this would add more of a real world element to it... it would still call mock gateway in the end.

----- Gateway -----
A blank method had to be included for the project to compile for a usecase test to run. Same method signature had to be included in the interface.

---- Usecase -----
Usecase has 2 things to do for this endpoint:
1) Call the thing that maps the data
2) Use that data to call the gateway that will insert it.
Reason for usecase calling the mapping the data is because we're using MadeTech Flavoured Clean Architecture. As per that architecture the Gateway should only be receiving the Domain object, not the request object. Therefore it was decided to put the call to mapping into the usecase.

Duslerke committed:
Added Gateway tests for POST initial process document endpoint

There should be an extra tests, as well as some changes made to the existing tests due to the recent change in specification.

Not doing all these things now, because doing the changes to match the specification will break existing tests, and I'd rather keep that to a separate commit for the sake of cleaner commits.

Duslerke committed:
Change of Specification changes + Visual changes

Specification changes:

1) According to the new specification the "processRef" (or "Id") now has to be of type "Guid" rather than "string", which it was before.

2) Also the "processType" now has to be an object with 2 fields (value, name) as opposed to string which it was before. This change was done to cheaply deal with potential process name changes in the future.

3) The main (initial) document body (corresponds to MatProcessData domain object) now has to have 2 extra fields: "ProcessDataAvailable" and "LinkedProcessId".

The goal of this commit was to change these things accross entire code base by changing all the method signatures, their implementations and their tests so that the already written tests would pass and there wouldn't be any type mismatches.

In addition to that the ordering of Assertion and the listing of properties and their arguments has been changed slightly to ensure consistency accross the codebase, so that finding parts of the code would be easier in the future.

There has also been some functional changes:
1) Tests that has [TestCase(...)] attribute have been changed to use "_faker" instead, to generate random Guids rather than to have them supplied through a testscase. Reason for this change was that [TestCase(...)] is known for not allowing any generation of values within it. Which means the guid would have to be pregenerated and stored somewhere else, which is far from clean. So instead these tests now use faker to generate a random guid. Triangulation is ensured not through the number of testcases, but instead through the shear inprobability (which is the essence of guid) of random guid matching whatever could be hardcoded within the test.

2) Change in all of the queries (some within test project, and one within actual implementation of gateway). It appears that there's some incompatibility between the MongoDB and .NET Guid type, when it comes to querying by it. Maybe it's possible to somehow configure MongoDB to avoid this, but I'm no expert in Mongo .NET drivers and their documentation is terrible. So I'm simply parsing the Guid to string within each query filter definition, which seems to make it work. A better change to it can be made later if needed.

3) Within the MatProcessData domain object definition, above the "Id" field, there has been an attribute: [BsonRepresentation(BsonType.ObjectId)] - not sure what it's for. It was compatible while "Id" was a string, however now that it's Guid, it's not - it gives an error message of: "ObjectId is not a valid representation for a GuidSerializer". Also I did some reading, and it seems that these 2 types differ in bit size as well: "Guid" is 128bits, while "ObjectId" is 96bits, so that would be lossy compression anyway - not sure if we would want that for our unique identifier of the document. So I just commented that attribute out for now.

At the point of this commit, all the tests pass except for the gateway's post method tests, due to that method not being implemented yet.

Duslerke committed:
PostInitialDocument gateway method basic implementation

Did just enough to get the tests pass. Tests only covered the insertion of the document, and not the try catch error handling, so there still is some work to do.

In the Second test for checking if it was just a single document inserted, there is this variable unclearedDocumentCount. It's there because it looks like the database doesn't get cleared after every test, so it makes it so that whatever records were inserted by other tests prior to this test, get to stay and ruin the day.
So you need to capture the initial document count.
Then I generate random amount of other documents for triangulation.
And only then I insert my actual document I'd want to insert IRL, and check if the count is 1 greater, than documents pre-inserted before test + documents pre-inserted before the main test's insertion.

The gateway method is no longer void since the specification changed from it giving no response to it giving just the processRef (Id)

Duslerke committed:
Change back "Id" from "Guid" to "string"

Changing back all tests, method signatures, variable types, variable generators and mongodb filter arguments to work with "Id" as a "string". Reason for the change is that there was a mistake in the specification. All tests passed.

Duslerke committed:
Made Post Usecase wrap up and return gateway output.

Like last time, one of the tests is impossible due to decision of sticking to 'static' factory rather than instance.

Added additional to test to cover the case of the usecase returning something. According to the new requirements, this endpoint is supposed to return the processRef (Identifier for the document) rather than empty response. Essentially 201 over 204. Inside the test had to resort to some trickery for returning something as simple as a string due mock acting in weird ways.

Changed few types in the PostInitialProcessDocumentResponse object, which was  unnecessarily added last time as it's only now that it's useful.

Implemented the return and wrapping logic in the usecase implementation file.

Duslerke committed:
TDD'ed basic PostInitialProcessDocument Controller method implementation

No error handling or validation yet - the format is still to be decided.

To proceed further with controller need to implement the validator class as well as some basic error handling and wrapping.

This commit contains the acceptance and unit tests for success cases as well as comments as a reminders for what other tests need to be in there once all other necessary pieces get implemented.

Change the controller return to ObjectResult. Reason for it is: while we decided that we want the 201 status code, the specification wants to return only one property of the object inserted rather than url to it + the object itself.

In processDataGatewayTests I removed the dev line left over from debugging.

In processDataGateway added a reminder to change "InsertOneAsync" to "InsertOne" as it was noticed that the async method does not detect duplicate inserts.

All test pass on my machine, however I don't have a way to test after having launched application up - due to missing mongo connection string??

Duslerke committed:
Added FluentValidation package reference

Also prepared the Validation folders both for validators and their tests.

Duslerke committed:
Wrote Validation tests for PostInitialProcessDocumentRequest object

For now it's just tests, no implementation, so CircleCI should fail.

It was decided to move away from internal validation (the [Required] attributes) because it did not provide error messages descriptive enough.

We used Fluent Validations as it's much easier to implement, use and test than building custom validator from scratch.

MirelaGeorgiou committed:
Added validation login

MirelaGeorgiou committed:
fixed tests, exception handling and validation

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataControllerAcceptanceTests.cs:
```diff
@@ -74,5 +77,48 @@ public void RunAfterAnyTests()
+            Assert.AreEqual(201, result.StatusCode);
+        }
+
+        //invalid request - needs validation setup first
```
Good to add these tests.

Duslerke replied:
Added the missing acceptance tests for the failing cases.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:
```diff
@@ -83,20 +82,71 @@ public void when_updateexistingprocessdocument_method_is_called_then_it_returns_
+
         [Test]
-        public void when_postinitialprocessdocument_method_is_called_then_it_returns_a_response_that_resource_was_created() //temporary test until actual implementation will be worked on.
+        public void when_postinitialprocessdocument__controller_method_is_called_then_it_returns_a_response_that_resource_was_created() //temporary test until actual implementation will be worked on.
```
I accidentally added double underscore

Duslerke replied:
Removed that

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:
```diff
@@ -83,20 +82,71 @@ public void when_updateexistingprocessdocument_method_is_called_then_it_returns_
+                ), Times.Once);
+        }
+
+        //Controlled calls validation test...
```
would be good idea to add, if possible

Duslerke replied:
Done.

Duslerke commented on mat-process-api.Tests/V1/Controllers/ProcessDataControllerTests.cs:
```diff
@@ -83,20 +82,71 @@ public void when_updateexistingprocessdocument_method_is_called_then_it_returns_
+            Assert.AreEqual(JsonConvert.SerializeObject(expectedResponse), JsonConvert.SerializeObject(resultContent));
+        }
+
+        //There should be test for given invalid request... essentially invalid GUID, but it should be done after the validation is done.
```
Since we'll do it in the acceptance tests. It would make sense to add it here as well.

Duslerke replied:
Done.

Duslerke commented on mat-process-api.Tests/V1/Gateways/ProcessesGatewayTests.cs:
```diff
@@ -66,21 +72,94 @@ public void test_that_gateway_return_object_matches_object_in_database()
+            processDataGateway.PostInitialProcessDocument(toBeInsertedDomainObject);
+
+            //assert
+            var startingDocumentCount = unclearedDocumentCount + toBeInsertedDocumentCount;
```
wrote this myself and find it confusing. should be unclearedDocumentCount + preInsertedDocumentCount (so that it wouldn't be confused with 'toBeInsertedDomainObject'.

Duslerke replied:
Done. These small changes will be squashed into single commit.

Duslerke commented on mat-process-api.Tests/V1/Gateways/ProcessesGatewayTests.cs:
```diff
@@ -66,21 +72,94 @@ public void test_that_gateway_return_object_matches_object_in_database()
+        }
+
+        [Test]
+        public void given_the_matProcessData_domain_object_when_postInitialProcessDocument_gateway_method_is_called_then_it_returns_id_of_the_inserted_document() //test that checks whether the db doesn't get cleared or overwritten somehow upon insertion
```
Forgot to remove the comment - it belongs with the test above.

Duslerke replied:
Removed.

Duslerke commented on mat-process-api.Tests/V1/Gateways/ProcessesGatewayTests.cs:
```diff
@@ -66,21 +72,94 @@ public void test_that_gateway_return_object_matches_object_in_database()
+            //act
+            string response_Id = processDataGateway.PostInitialProcessDocument(domainObject);
+
+            Assert.Throws<ConflictException>(() => processDataGateway.PostInitialProcessDocument(domainObject));
```
I'll add the comment saying that that's where the second insert happens. So the question wouldn't need to arise.

Duslerke replied:
Done. Though this would be nice, considering I myself had that question.

Duslerke commented on mat-process-api.Tests/V1/Helper/MatProcessDataHelper.cs:
```diff
@@ -17,17 +18,38 @@ public static MatProcessData CreateProcessDataObject()
-                PreProcessData = {},
+                ProcessDataAvailable = false,
+                ProcessDataSchemaVersion = faker.Random.Int(0,10),
                 ProcessStage = faker.Random.Word(),
```
Might not need to change this, but need to take a look into it anyway. According to new requirement, 'ProcessStage' is supposed to always be 'Not Completed' upon initial insert of the document. Have forgotten how this method is used.. if it's for initial or default MatProcessData domain object, then it will need to be changed. If it's for simulating already inserted docs, then it's fine as is.

Duslerke replied:
The way it has been used doesn not matter in the context of ProcessStage. All that's being checked is whether the inserted docs data matches the retrieved one's. Therefore I can change this helper method to generate the "Not Completed" value for ProcessStage rather than the random value. It's better this way for consistency reasons, because the method in the edit is the method for generating **initial** document as can be seen from from other properties it generates.

Duslerke commented on mat-process-api.Tests/V1/UseCase/ProcessDataUseCaseTests.cs:
```diff
@@ -83,5 +83,52 @@ public void check_gateway_returns_expected_response()
+            //assert
+            mockMatGateway.Verify(g => g.PostInitialProcessDocument(It.IsAny<MatProcessData>()), Times.Once);
+
+            // I need to include into assertion just enough object properties so that it would be demonstrated that all the data from the request is in a mapped-to object (probably 1 would be enough).
```
Need at least one assertion so that it would be clear that gateway is being called with that output of the factory method. Currently there's no way to tell. The only thing test whos is that the gateway is called with an object with the correct type, but it should also show that the data is not 'null' inside.

Duslerke replied:
Looks there's a better way. In the Verify bit you can check whether the mock object is called with the correct properties of the passed in object. Unfortunately checking the object itself does not work due to C# checking whether object references match instead of whether object contents match.

The call to factory in the UseCase gives ref1, while the call in the test (domainObject) gives ref2, hence test fails, if you don't check the contents of the properties themselves.

Duslerke commented on mat-process-api.Tests/V1/Helper/MatProcessDataHelper.cs:
```diff
@@ -17,17 +18,38 @@ public static MatProcessData CreateProcessDataObject()
                 DateCreated = faker.Date.Recent(),
                 DateLastModified = faker.Date.Recent(),
-                ProcessData = {},
-                ProcessDataSchemaVersion = faker.Random.Int(0,10),
```
Combined with "Validators/PostInitialProcessDocumentRequestValidatorTests.cs" Line 39. This has a probability of 10% of failing any controller test that uses this helper. Due to validation fire-ing off upon SchemaVersion = 0, when it's not intended (as judging by test's purpose). The starting Schema version will be '1', so need to exclude zero anyway.

Duslerke replied:
Changed it to generate numbers from 1 to 10 instead, not to hit zero.

Duslerke commented on mat-process-api.Tests/V1/Validators/PostInitialProcessDocumentRequestValidatorTests.cs:
```diff
@@ -0,0 +1,102 @@
+            _postInitProcessDocValidator.ShouldHaveValidationErrorFor(r => r.processRef, request);
+        }
+
+        //official website - basic setup of validator with an example of how to use it  https://fluentvalidation.net/start#creating-your-first-validator
```
Remove the handover comments.

Duslerke replied:
Done. They're no longer needed as the implementation has already been done.

Duslerke commented on mat-process-api/V1/Boundary/PostInitialProcessDocumentRequest.cs:
```diff
@@ -0,0 +1,21 @@
+{
+    public class PostInitialProcessDocumentRequest
+    {
+        [Required] public string processRef { get; set; }
```
The [Required] ComponentModel DataAnnotations are not needed anymore, because this object's validator takes care of it now.

Duslerke replied:
Removed. Reasoning behind the removal is that ... if  for some reason any of these properties would not need to be Required anymore... you'd have to change it in the validations (where you expect) and in the here (where you wouldn't) - it's 2 places doing same thing - not very good anyway.
Also if anyone were to need to add another property, this would cause confusion due to there being 2 places that do this, yet this one being inactive.

Duslerke commented on mat-process-api/V1/Gateways/ProcessDataGateway.cs:
```diff
@@ -29,5 +30,33 @@ public MatProcessData GetProcessData(string processRef)
+            }
+            catch(AggregateException ex)
+            {
+                ex.Handle((x) =>
```
Not sure what's going on here... need some comments to clarify this.

Duslerke replied:
Added a comment explaining why this is needed.

Duslerke commented on mat-process-api/V1/Validators/PostInitialProcessDocumentRequestValidator.cs:
```diff
@@ -0,0 +1,33 @@
+    {
+        public PostInitialProcessDocumentRequestValidator()
+        {
+            ValidatorOptions.CascadeMode = CascadeMode.StopOnFirstFailure;
```
Not sure what this is here for. Need to ask and document in the comment. Why can't we return all the errors at once?

Duslerke replied:
Added comment clarifying that. It's different from what you'd expect by just looking at it:
It's used to stop the validator from going deeper into any Rule after first failure. 

It doesn't stop checks on each Rule, it's stops continuous checks within the failing Rule. Needed because it stops from checking the properties of the object, if the check for object's existence has failed.

Duslerke commented on mat-process-api/V1/Validators/PostInitialProcessDocumentRequestValidator.cs:
```diff
@@ -0,0 +1,33 @@
+                .DependentRules(() =>
+                {
+                    RuleFor(postReq => postReq.processType.value).NotEmpty().WithMessage("Process type value must be provided.")
+                     .NotEmpty().WithMessage("Process type value must be provided."); ;
```
Duplicate Line, double semicolon. Might need to change one of them into "NotNull". Also, I think the WithMessage can be trimmed into a single call after both validations.

Duslerke replied:
Fixed that. Though apparently the 'null' check might be pointless. Unless someone decides to make it into 'int?' in the future.

Duslerke commented on mat-process-api/V1/Validators/PostInitialProcessDocumentRequestValidator.cs:
```diff
@@ -0,0 +1,33 @@
+                    RuleFor(postReq => postReq.processType.value).NotEmpty().WithMessage("Process type value must be provided.")
+                     .NotEmpty().WithMessage("Process type value must be provided."); ;
+                    RuleFor(postReq => postReq.processType.name).NotNull().WithMessage("Process type name must be provided.")
+                    .NotEmpty().WithMessage("Process type name must be provided.");
```
Same .WithMessage(..) twice. Might be able to trim it to single call after both validations.

Duslerke reviewed (changes requested):
Need these changes.

Duslerke committed:
Added missing Acceptance Tests that cover the failure cases.

400 acceptance test - the only thing I could do with the expected response is, to simply see if what I wrapped up manually (in the arrange phase), matches the automatic wrapping im the controller, hence this frankenstein appears: '((IList<ValidationFailure>)expectedResponse.Value)[0].ErrorMessage' in the Assert statement.

409 acceptance test doesn't test the error message match, because the error message depends on random value generated + it requires me to capture it when error is thrown. This can't be done due to the try catch block.

I could hardcode the message and make the guid change, but that would look terrible. So instead, I check whether anything gets appended to the prewritten string. If the conflict error is thrown, but there's nothing appended, then clearly, something has gone wrong in the error handling.

Duslerke committed:
Fixed bug - empty {error message} part of the exception string.

A test reveals a bug, where the InnerException has not been passed into custom exception, hence information loss, and incomplete custom error message.

The fix, create a constructor for custom exception that takes inner exception as an argument. Which allows to pass the previously missing information in.

The custom exception should probably be extracted away into some exceptions folder, but that's probably not for now.

Duslerke committed:
Removed unneeded [Required] annotations. Renamed variable for more clarity. Removed some typos, unneeded comments. Added some needed comments.

Changed 'toBeInsertedDocumentCount' into preInsertedDocumentCount, so that it wouldn't be confused with 'toBeInsertedDomainObject', which is the start of the test.

Reasoning behind the [Required] removal is that ... if  for some reason any of these properties would not need to be Required anymore... you'd have to change it in the validations (where you expect) and in the here (where you wouldn't) - it's 2 places doing same thing - not very good anyway.
Also if anyone were to need to add another property, this would cause confusion due to there being 2 places that do this, yet this one being inactive.
Tests still pass as they should.

Duslerke commented:
The unresolved ones, I'll need to take a closer look tomorrow, they require some investigation.

Duslerke committed:
Fixed few issues regarding MatProcessData generation in the Helper method. Also Added few comments in the Validator and Gateway explaining few less clear parts of the code.

----------------- Validator -----------------
Fixed the duplicate Rule for 'ProcessType.value'. It had '.NotEmpty()' twice, so I changed one of them into '.NotNull', Which might become useful if integer becomes nullable, or gets set to null in the request somehow.

Added the comment explaining the CascadeMode validator option.

----------------- Gateway -----------------
Added a comment explaining the AggregateException handling.

----------------- Helper -----------------
The 'ProcessDataSchemaVersion' random value between (0,10) combined with the not zero test in the validation itself, has a probability of 10% of failing any controller test that uses this helper. There aren't any yet, but as a good precaution, it's best to do changes before issue happens.
The starting Schema version will be '1', so need to exclude zero anyway, if not for the issue stated, then for consistency.

The way MatProcessData object has been used does not matter in the context of ProcessStage. All that's being checked is whether the inserted docs data matches the retrieved one's. Therefore I can change this helper method to generate the "Not Completed" value for ProcessStage rather than the random value. It's better this way for consistency reasons, because the method in the edit is the method for generating **initial** document as can be seen from from other properties it generates.

Duslerke committed:
Completed the incomplete Gateway test that checks whether the Usecase calls the gateway with the correct data.

This test is supposed to check whether the Usecase calls the Gateway with whatever ProcessDataFactory converter (from Request obj to Domain obj) would output.

However, previously the test was only checking whether the Usecase calls the Gateway with the correct type of the object. This excluded the check whether the data was correct.

So in this commit, I add a check for the Data contained in the object as well:

In the Verify bit you can check whether the mock object is called with the correct properties of the passed in object. Unfortunately checking the object itself (which would be a shortcut to typing every property check) does not work due to C# checking whether object references match rather than whether object contents match:
The call to factory in the UseCase gives ref1, while the call in the test (domainObject) gives ref2, hence test fails, if you don't check the contents of the properties themselves instead.

Duslerke committed:
Added Post Initial Document Request Validator into dependency injection setup. And did the compatability changes.

Create an interface for the PostInitialDocumentRequestValidator.

Made that validator an instance of that interface. Made an interface require '.Validate()' method, as it's the thing that's being used from within the controller.

In the controller I added validator as a parameter, rather than having it set to instance in the constructor.

In the tests, in order to match new definition of controller, I added mocks to deal with that. Where needed I set mocks to return a no-validation-error result, because all the tests at the time were of success cases only.

Duslerke committed:
Tests name change to match the fact that it's no longer unique situation.

Changed the test name to denote that this is the test for success case, with the success input. As I'm about to add failure cases.

Duslerke committed:
Added a test that checks if PostInitDoc controller method calls the validator with correct data.

This tests checks whether the controller calls the Validator and passes the request into the Validator.

Duslerke committed:
Added Exception Handling unit tests for the PostInitDoc controller method.

Difference between these and the acceptance ones is that these ones use mocks, rahter than actual instances.

MirelaGeorgiou commented on mat-process-api.Tests/V1/Factories/ProcessDataFactoryTest.cs:
```diff
@@ -45,16 +47,51 @@ public void can_create_mat_process_data_object_from_populated_object()
+            Assert.AreEqual("Not completed", domainObject.ProcessStage); //it's not confirmed yet whether it's going to be int or string
+            Assert.Null(domainObject.LinkedProcessId);
+
+            // Assert.AreEqual(new { }, domainObject.PreProcessData); //Causes error --> Expected: <{ }> (<>f__AnonymousType0); But was:  <{ }> (<> f__AnonymousType0)
```
we will need to clean up commented code

Duslerke replied:
alright.
